### PR TITLE
gh-80480: Emit DeprecationWarning for array's 'u' type code

### PR DESCRIPTION
**Pull request by hugovk**
The "u" type code for `array` was deprecated in Python 3.3 but only via documentation: https://docs.python.org/3.10/library/array.html

Let's add the `DeprecationWarning`.

It was originally planned to be removed in Python 4.0, but that was before we knew if 3.10 or 4.0 would come after 3.9:

> '4.0' is a stand-in for 'sometime after 2.7.final', scheduled for Jan 2020.

https://github.com/python/cpython/issues/80480#issuecomment-1093818167

So as this is going into ~3.12~ 3.13, let's target ~3.14~ ~3.15~ 3.16 for removal.

* Issue: gh-80480

### CODE REVIEW

**Review by methane — APPROVED** [review]

File: Lib/test/test_array.py [hunk 1]
```diff
@@ -1205,6 +1216,15 @@ def test_issue17223(self):
         self.assertRaises(ValueError, a.tounicode)
         self.assertRaises(ValueError, str, a)
 
+    def test_typecode_u_deprecation(self):
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always", DeprecationWarning)
+            array.array("u")
+        self.assertGreaterEqual(len(w), 1)
+        for warning in w:
+            self.assertIs(warning.category, DeprecationWarning)
```
Comment (ezio-melotti) [inline Lib/test/test_array.py]: Can't you use [`assertWarns`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarns) here?

Comment (hugovk) [reply Lib/test/test_array.py]: Ah yes, much better!

**Comment by asedeno** [comment]
As a user of the `'u'` array type, I'd like to see this postponed until a UCS4 `'w'` array type is in place to migrate to, as was proposed in https://github.com/python/cpython/issues/80480#issuecomment-1093818179.

**Comment by hugovk** [comment]
Thanks for the ping. I commented in the issue, let's continue there.

**Comment by arhadthedev** [comment]
Do you have plans to include this deprecation into 3.12b1?

**Comment by hugovk** [comment]
Updated to target removal in 3.16, following addition of `'w'` in https://github.com/python/cpython/pull/105242 and following the plan at https://github.com/python/cpython/issues/80480#issuecomment-1541327809.

File: Lib/test/test_re.py [hunk 2]
```diff
@@ -1522,10 +1522,11 @@ def test_bug_6561(self):
         for x in not_decimal_digits:
             self.assertIsNone(re.match(r'^\d$', x))
 
+    @warnings_helper.ignore_warnings(category=DeprecationWarning)  # gh-80480 array('u')
     def test_empty_array(self):
         # SF buf 1647541
         import array
-        for typecode in 'bBuhHiIlLfd':
+        for typecode in 'bBhuHiIlLfd':
```
Comment (methane) [inline Lib/test/test_re.py]: Would you add 'w' here?

File: Doc/whatsnew/3.12.rst [hunk 3]
```diff
@@ -1104,6 +1104,8 @@ Pending Removal in Python 3.14
 
   (Contributed by Jason R. Coombs and Hugo van Kemenade in :gh:`93963`.)
 
+APIs:
+
```
Comment (methane) [inline Doc/whatsnew/3.12.rst]: Maybe, another pull request?

File: Doc/whatsnew/3.13.rst [hunk 4]
```diff
@@ -122,11 +122,21 @@ Optimizations
...
+Pending Removal in Python 3.15
+------------------------------
+
 * :mod:`wave`: Deprecate the ``getmark()``, ``setmark()`` and ``getmarkers()``
   methods of the :class:`wave.Wave_read` and :class:`wave.Wave_write` classes.
   They will be removed in Python 3.15.
   (Contributed by Victor Stinner in :gh:`105096`.)
 
+Pending Removal in Python 3.16
+------------------------------
```
Comment (methane) [inline Doc/whatsnew/3.13.rst]: I don't think this sections are needed.

File: Lib/test/test_re.py (hunk 2 above)
Comment (hugovk) [reply Lib/test/test_re.py]: Added.

File: Doc/whatsnew/3.12.rst (hunk 3 above)
Comment (hugovk) [reply Doc/whatsnew/3.12.rst]: Removed.

File: Doc/whatsnew/3.13.rst (hunk 4 above)
Comment (hugovk) [reply Doc/whatsnew/3.13.rst]: We have these for other releases:

* https://docs.python.org/3.13/whatsnew/3.11.html#pending-removal-in-python-3-12
* https://docs.python.org/3.13/whatsnew/3.12.html#pending-removal-in-python-3-13
* https://docs.python.org/3.13/whatsnew/3.12.html#pending-removal-in-python-3-14
* https://docs.python.org/3.13/whatsnew/3.12.html#pending-removal-in-future-versions

They help users plan when to update their code, and also help us later when the time comes to remove old code.

Comment (methane) [reply Doc/whatsnew/3.13.rst]: OK. I have seen only 3.11.html. It only has "Pending removal in Python 3.12" (3.11+1) section and only "have been deprecated in earlier Python releases" are listed.

But now I see 3.12.html and you are right.

File: Doc/whatsnew/3.13.rst [hunk 5]
```diff
@@ -122,11 +122,21 @@ Optimizations
...
+
+* :mod:`array`'s ``'u'`` format code, deprecated in docs since Python 3.3,
+  emits :exc:`DeprecationWarning` since 3.13
+  and will be removed in Python 3.16.
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst:132]: Can you suggest a replacement? I suppose that the ``'w'`` format should be used instead.

File: Modules/arraymodule.c [hunk 6]
```diff
@@ -2674,6 +2674,15 @@ array_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
...
+    if (c == 'u') {
+        if (PyErr_WarnEx(PyExc_DeprecationWarning,
+                         "The 'u' type code is deprecated and "
+                         "will be removed in Python 3.16",
+                         1)) {
+            return NULL;
+        }
+    }
+
     if (PySys_Audit("array.__new__", "CO",
```
Comment (vstinner) [inline Modules/arraymodule.c]: I suggest to move the warning after the audit.

File: Doc/whatsnew/3.12.rst [hunk 7]
```diff
@@ -1113,7 +1113,7 @@ Pending Removal in Python 3.14
   :exc:`DeprecationWarning` to be emitted when it is used.
 
 * Creating immutable types (:data:`Py_TPFLAGS_IMMUTABLETYPE`) with mutable
-  bases using the C API.
+  bases using the C API  (:gh:`95388`).
```
Comment (vstinner) [inline Doc/whatsnew/3.12.rst]: The change looks unrelated, you might make it separately, so you can backport it to 3.12 branch.

File: Lib/test/test_buffer.py [hunk 8]
```diff
@@ -3183,6 +3184,7 @@ def cmptest(testcase, a, b, m, singleitem):
                             self.assertEqual(m.tobytes(), a.tobytes())
                             cmptest(self, a, b, m, singleitem)
 
+    @warnings_helper.ignore_warnings(category=DeprecationWarning)  # gh-80480 array('u')
     def test_memoryview_compare_special_cases(self):
```
Comment (vstinner) [inline Lib/test/test_buffer.py]: Would it be possible to move the "u" test in a new method?

File: Doc/whatsnew/3.12.rst (hunk 7 above)
Comment (hugovk) [reply Doc/whatsnew/3.12.rst]: Good idea, removed.

File: Lib/test/test_buffer.py (hunk 8 above)
Comment (hugovk) [reply Lib/test/test_buffer.py]: Yep, now moved to its own method. Makes it easier to remove when the time comes.

File: Doc/whatsnew/3.13.rst [hunk 9]
```diff
@@ -122,11 +122,22 @@ Optimizations
 Deprecated
 ==========
 
+Pending Removal in Python 3.15
+------------------------------
+
 * :mod:`wave`: Deprecate the ``getmark()``, ``setmark()`` and ``getmarkers()``
   methods of the :class:`wave.Wave_read` and :class:`wave.Wave_write` classes.
   They will be removed in Python 3.15.
```
Comment (vstinner) [inline Doc/whatsnew/3.13.rst:127]: This sentence now sounds redundant with the section title. Same remark with "will be removed in Python 3.16" below.

**Review by vstinner — APPROVED** [review]
LGTM. I just left minor remarks on the doc. It's up to you to address them or ignore them ;-) Thanks for the updates.

File: Doc/whatsnew/3.13.rst (hunk 9 above)
Comment (hugovk) [reply Doc/whatsnew/3.13.rst:127]: I'll instead remove these section titles, I just realised they're missing this intro text (emphasis added):

> The following Python APIs have been _deprecated in earlier Python releases_, and will be removed in Python 3.12.

For example:

* https://docs.python.org/3.13/whatsnew/3.11.html#pending-removal-in-python-3-12
* https://docs.python.org/3.13/whatsnew/3.12.html#pending-removal-in-python-3-13
* https://docs.python.org/3.13/whatsnew/3.12.html#pending-removal-in-future-versions
